[selectors] Add focus-visible-013.html test and fix focus-visible-007.html

mrego opened the pull request:
This test had wrong assertions and it was too complex,
so this patch is simplifying it.

A part of this test is extracted to a separated test
focus-visible-013.html, where we check that :focus-visible
does not match after mouse click even if previous focused element
was matching :focus-visible.

And then we simplify the focus-visible-007.html test, to check that
an active element that doesn't match :focus-visible,
will start matching :focus-visible after after a keyboard interaction
that doesn't move the focus (in this case the "Shift" key).

mrego committed:
[selectors] Add focus-visible-013.html test

This test is extracted from focus-visible-007.html
that has other problems.

In focus-visible-013.html we check that :focus-visible
does not match after mouse click even if previous focused element
was matching :focus-visible.

mrego commented:
This new `focus-visible-007.html` passes in Chromium and fails in Firefox.

This test is checking this part of the spec:
> * If the user interacts with the page via the keyboard, the currently focused element should match `:focus-visible` (i.e. keyboard usage may change whether this pseudo-class matches even if it doesn’t affect `:focus`).

So there's an element that was focused via mouse without triggering `:focus-visible` and then the user press the `Shift` key. And even when the focus doesn't move, the element now matches `:focus-visible`.

Note that in Chromium this doesn't happen for all keyboard interactions, as `focus-visible-012.html` is explicitly checking that using a modifier key (like `Ctrl`) doesn't trigger `:focus-visible`.

Firefox is not doing this for any keyboard interaction as far as I've tested.

CC @alice

Loirooriol commented on css/selectors/focus-visible-007.html:
```diff
@@ -10,98 +10,48 @@
   <style>
-    [data-hadkeydown] :focus-visible {
-      outline: darkgreen solid 5px;
+    @supports not (selector(:focus-visible)) {
```
I think you can remove the `@supports` and just use

```css
      :focus {
        outline: red solid 5px;
      }
```
since, if `:focus-visible` is supported, the `:focus-visible` and `:focus:not(:focus-visible)` below will have more precedence.

mrego replied:
Thanks for the review.

True thing, I could remove that. But in that case I'd probably like to add a comment about "This is to detect a problem in the tests for browsers that don't support :focus-visible". Which is already quite clear with the `@supports`.

Also most tests on this suite use that, maybe we can do a follow-up patch to change that in more places if you think it's really important.

Loirooriol replied:
Fine, not really important

Loirooriol commented on css/selectors/focus-visible-007.html:
```diff
@@ -10,98 +10,48 @@
-        assert_true(false, "send_keys not implemented yet");
+      target.addEventListener("keyup", t.step_func(function() {
+        assert_equals(getComputedStyle(target).outlineColor, "rgb(0, 128, 0)", `outlineColor for ${target.tagName}#${target.id} should be green`);
         t.done();
```
Nit: I think you can replace the `t.step_func` with `t.step_func_done` and then remove this `t.done()`.

mrego replied:
Done.

mrego commented on css/selectors/focus-visible-007.html:
```diff
@@ -10,98 +10,48 @@
       }));
-    }, "Using keyboard while element is focused should trigger :focus-visible; using mouse without moving focus should not cancel it; moving focus using mouse should cancel it.");
+
+      test_driver.click(target).then(() => done());
```
I can remove this, in theory this won't be needed, but it looks like Chromium has some kind of leak bug due to this stuff: https://crbug.com1164600

Anyway once the modified test is imported in Chromium, I'll verify there if there's any issue, and propose a change to test according to that.

mrego commented on css/selectors/focus-visible-013.html:
```diff
@@ -0,0 +1,52 @@
+  async_test(function(t) {
+    initial.addEventListener("focus", t.step_func(function() {
+      assert_equals(getComputedStyle(initial).outlineColor, "rgb(0, 128, 0)", `outlineColor for ${initial.tagName}#${initial.id} should be green`);
+      test_driver.click(target).then(() => t.done());
```
I've changed this and used `step_func_done` as you suggest, I think now the test is clearer.

Loirooriol commented on css/selectors/focus-visible-013.html:
```diff
@@ -0,0 +1,52 @@
+  async_test(function(t) {
+    initial.addEventListener("focus", t.step_func(function() {
+      assert_equals(getComputedStyle(initial).outlineColor, "rgb(0, 128, 0)", `outlineColor for ${initial.tagName}#${initial.id} should be green`);
+      test_driver.click(target).then();
```
Nit: you can remove the empty `.then()`

mrego replied:
Oops, sorry.

Thanks for the review.

alice commented:
The opening comment says the test had some "wrong assertions", but doesn't say what they are, and the change is too complex to figure it out from the diff.

What were the wrong assertions?

alice commented on css/selectors/focus-visible-007.html:
```diff
@@ -10,98 +10,47 @@
   <ol id="instructions">
     <li>If the user-agent does not claim to support the <code>:focus-visible</code> pseudo-class then SKIP this test.</li>
-    <li>Use the mouse to focus the element below that says "Click me first."</li>
-    <li>If the element has a red outline, then the test result is FAILURE.</li>
```
Why change this?

mrego replied:
I believe the test was kind of complex and testing 2 different things, that's why I split it in 2 different tests.
Then I don't need so many steps and elements on this first test, and that's why I removed this part.

alice replied:
You changed this from a failure to a success, but the test doesn't pass unless both steps (initial focus from click not matching :focus-visible, and later pressing a key triggering :focus-visible matching) pass. So it's misleading to claim that the test is a success after step one. Why not leave this specific text the way it was?

mrego replied:
Sorry my previous comment was too quick and wrong.

The steps on the new version of the test are:
  <ol id="instructions">
    <li>If the user-agent does not claim to support the <code>:focus-visible</code> pseudo-class then SKIP this test.</li>
    <li>Use the mouse to focus the element below that says "Click me."</li>
    <li>If the element has a blue outline, then the test result is a SUCCESS.</li>
    <li>Press the SHIFT key.</li>
    <li>If the element now has green outline, then the test result is SUCCESS.</li>
  </ol>

The first time you click only `:focus` matches and not `:focus-visible`, and it has a blue outline.
Then after clicking SHIFT, `:focus-visible` matches and the outline changes to green.

I can come back to the previous colors I guess, but changing the test again. The thing was that I was having a hard time understanding the whole test and I believed the new version was better to understand (apart from the issues with the wrong assertions).

alice replied:
I agree with splitting the test into two parts, apologies for not being clear about that.

Splitting this out into two tests means that the later part of the manual steps (click on the second element) are no longer needed, I agree with that.

However, previously the manual test made it clear that not passing the first part of the test (clicking the element shouldn't trigger :focus-visible matching) is an instant fail, meaning the second part of the test (pressing shift) is irrelevant. That's why it had the mechanism for setting up that red border to demonstrate that.

Can we have separate changes for splitting this out into two tests, and simplifying focus-visible-007.html? That might make this review easier for both of us.

mrego replied:
Ok, maybe I went too far away changing this test, I know understand your point. Sorry for all the back and forth.

I can separate this in different PRs, or maybe simplify this PRs by just removing the last steps in `focus-visible-007.html` but keep the rest (fixing the wrong assertions), and then add the proper manual instructions to `focus-visible-013.html`. Sorry for the mess.

alice replied:
Thank you for your patience with all my comments! Even just splitting this change into two separate commits for the two different changes would help, or otherwise just keep this change as the splitting out (uncontroversial and we can land it right away), and we can work on the simplification/leak fixing etc. separately.

mrego replied:
So I've tried to just split this in two tests, but `focus-visible-007.html` has other problems too.

The test starts by sending a SHIFT key, then it sends the SHIFT key again after focus, and at that point it starts to test things. I believe I can fix `focus-visible-007.html` in a separated PR and keep this PR for adding a new test that covers part of `focus-visible-007.html` that I'll get rid of later.

mrego replied:
So I've finally done 2 different commits.
One of them adds a new `focus-visible-013.html` test, which is the same than before but with instructions to run it manually.
And then another commit fixes `focus-visible-007.html`, keeping most of its current design, but removing the wrong parts and the extra checks that are now tested in `focus-visible-013.html`.

alice replied:
Thank you so much for this extra work! This cleanup looks excellent.

alice commented on css/selectors/focus-visible-013.html:
```diff
@@ -0,0 +1,52 @@
+
+<div id="initial" tabindex="0">Initial</div>
+<div id="target" tabindex="0">Target</div>
+
```
Can we have some manual testing instructions as well?

mrego replied:
These tests are automatic, so I'm not sure why we need the manual testing instructions. I can add them though.

alice replied:
I use the manual testing version when I am debugging browser code during implementation, and I figure others might like to as well.

mrego replied:
Ok, I'll add them.

mrego commented on css/selectors/focus-visible-007.html:
```diff
@@ -10,98 +10,47 @@
-        });
-
-        const test_modality_change = t.step_func(() => {
-          assert_equals(getComputedStyle(one).outlineColor, "rgb(59, 153, 252)");
```
These asserts that were checking "weird" colors are the wrong assertions.

alice replied:
That's odd, they would have passed when I added them. When did they begin failing?

mrego replied:
It looks like this is somehow failing since long time ago, when `test_driver.send_keys()` was implemented in Chromium: https://chromium-review.googlesource.com/c/chromium/src/+/1652635

alice replied:
Looking at https://crbug.com/976438, I see the original version of this test is also triggering a leak. I'm not sure whether that's a problem with the test, or the code under test.

mrego replied:
That leak is due to `test_driver.click()` that is causing a leak on Chromium unless we wait for the promise to resolve.

The Chromium bug is https://bugs.chromium.org/p/chromium/issues/detail?id=1164600

I've fixed similar leaks in focus-visible tests in https://bugs.chromium.org/p/chromium/issues/detail?id=1163172

And once the fixed version of `focus-visible-007` gets imported into Chromium, I'll check if it has leaks or not there, and fix them in a similar way.

alice replied:
Wonderful, thanks for the investigation and explanation!

mrego reviewed (commented):
I've replied inline in one of the wrong assertions checking weird colors, not the ones defined on the CSS of the test.

Apart from that you can see the final status of the test in https://github.com/web-platform-tests/wpt/blob/638eb0680fcac43d587eea75e26733e2edb741b3/css/selectors/focus-visible-007.html, so it's easier to understand.

mrego committed:
[selectors] Fix focus-visible-007.html

The test had a bunch of issues that are fixed by this patch.
Also a part of this tests has been extracted to focus-visible-013.html
which allows us to further simplify it.

alice commented on css/selectors/focus-visible-013.html:
```diff
@@ -0,0 +1,61 @@
+<ol id="instructions">
+  <li>If the user-agent does not claim to support the <code>:focus-visible</code> pseudo-class then SKIP this test.</li>
+  <li>Press the TAB key.</li>
+  <li>If the element that says "Initial" has a green outline, then the test rsult is a SUCCESS.</li>
```
I would still prefer not to say the test is a success until we have actually tested the critical thing (i.e. that moving focus using the mouse from an element which matched `:focus-visible` causes the newly focused element *not* to match `:focus-visible` - and yes, this was a real bug that we temporarily had during our implementation process).

I don't want to block on it since this has dragged on so long, but I would prefer it if we could go back to the original behaviour/wording stating when the test is a FAILURE (analogous to an `assert()`), instead of having two SUCCESS states.

mrego replied:
Ok, I understand. I'll modify the instructions so SUCCESS is only at the end, but also explain intermediate FAILURES there. Thanks.

mrego commented:
Thanks for the review, uploaded new version updating the instructions for `focus-visible-013` and will merge it.
